Overlay multiple schema files

gcheadle-vmware opened the pull request.

Commit:
feat: Support overlaying multiple schema files in root library

- Removed dvpp reciever on overlay function in workspace package

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>
Signed-off-by: Cari Dean <cdean@pivotal.io>

gcheadle-vmware committed:
feat: Overlay schema files from root scoped to private library

- Add lib reference to Document Schema
- Create plumbing to pass along library targeted schemas
- Failing test for 'with_schema()' library fucntion

Signed-off-by: Steven Locke <slocke@pivotal.io>

Commit:
feat: Fix incorrect annotation in error message

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

cari-lynn committed:
feat: Add with_schema() to library module

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

cari-lynn committed:
Refactor: separate schema logic from dataValues in library module

- follow pattern from root library to collect schema docs and
  datavalues docs separately

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware committed:
Refactor: separate method and remove uneeded variables.

Signed-off-by: Cari Dean <cdean@pivotal.io>

cari-lynn commented:
Remaining items before the PR is ready to be reviewed:
- [x] extract duplicate code for dv and schema into another package
- [x] rename either package libraryref or type libraryref
- [x] Rename the types that were extracted from common dv and schema code
- [x] Does the 'common' dv and schema code belong in `template` package?
- [x] since we can use data values in text templates, we should reconsider not passing schema into `template_loader.go` for evalText and evalStarlark
- [x] 🆕  Support `#@ library/ref` annotation on schemas docs. 
  - [x] add tests
  - [x] if no schema file exists in a library, use the one annotated with `library/ref` as the base data values file.
- [x] Address TODO's from previous bullet point working on `library/ref`.
- [x] should the duplicated `UsedInLibrary` function have similar behaviour? Currently they do different things
- [x] in `library_loader.go` Values() takes “overlays” because of command-line flags. Does Schemas() need to take “overlays”?
- [x] does it make sense to pass in `nil` to `schemas()` function in root library? Or is there a better way to do that?
- [x] Do we need to have two interfaces for schema and for data values?

cari-lynn committed:
Refactor: dried up library references for dv and schemas

- moved library ref to its own package inside of workspace
- introduce ExtractLibRefs interfaces to extract library references from
  data values and schemas

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>
Signed-off-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/cmd/template/schema_test.go:
```diff
@@ -933,6 +933,162 @@ foo: #@ data.values.foo
+
+		schemaYAML2 := `#@ load("@ytt:overlay", "overlay")
+#@schema/match data_values=True
+#@overlay/match by=overlay.all
```
something to think about... why do we need `#@overlay/match by=overlay.all` on schemas but not on data/values?

cari-lynn replied:
Good catch.. we don't need this annotation. This must've been a copy/paste mistake. Fixed.

cppforlife commented on pkg/schema/schema.go:
```diff
@@ -22,6 +22,9 @@ type DocumentSchema struct {
 	defaultDVs *yamlmeta.Document
 	Allowed    *DocumentType
+
+	originalLibRef []template.LibRefPiece
+	libRef         []template.LibRefPiece
```
is LibRefPiece really related to templating?

pivotaljohn replied:
This also came up in an in-person review we did, yesterday.  👍🏼

Finding/creating the right home for this (and related bits) is on the to-do list, in this issue/PR.

DennisDenuto replied:
We agreed it felt 'off' being in template. We landed on `workspace/ref`.  There was sufficient logic in how libref's are parsed / handled.  We treat this new 'ref' package as a 'go' library used by schemas and datavalues.

DennisDenuto replied:
resolving since this change is pretty outdated. `LibRefPiece` was also renamed to `LibRef`

cari-lynn committed:
feat: check schema when text templating uses libraries

- uses schema overlays for libraries when evaluating the libraries
  inside text templating or Starklark files

cari-lynn committed:
refactor: move and reword schema tests

Signed-off-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Bug fix: Schemas referencing nested libraries are correctly used

- A schema referencing a nested library, did not have its default data
values applied correctly
- A schema referencing a nested library (2 levels in depth) was not
having its type check enforced

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Invert if statements to have true case first
- Use 'HasLibRef' function instead of checking if libref is 0 length.
The HasLibRef function has a comment explaining its purpose

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: Provide error to user when schema isn't used

- Introduce similar 'used' logic as data values but for schemas

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- inline the overlay function to its respective usages

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Improve unused doc error message

Error message around a schema/datavalue not being used now
differentiates between the two document types.

Authored-by: Dennis Leon <leonde@vmware.com>

pivotaljohn reviewed (approved):
LGTM!
